Shared/ViewModels/DevicesViewModel.swift

```diff
@@ -0,0 +1,233 @@
+
+            backgroundStates.append(.deletingAllDevices)
+
+            deviceTask = Task { [weak self] in
+                do {
+                    try await self?.deleteAllDevices()
+                    await MainActor.run {
+                        self?.state = .content
+                    }
+                } catch {
+                    guard let self else { return }
+                    await MainActor.run {
+                        self.state = .error(.init(error.localizedDescription))
+                    }
+                }
+
+                await MainActor.run {
+                    self?.backgroundStates.remove(.deletingAllDevices)
+                }
+            }
+            .asAnyCancellable()
+
+            return state
+        }
+    }
+
+    // MARK: - Load Devices
+
+    private func loadDevices() async throws {
+        let request = Paths.getDevices()
+        let response = try await userSession.client.send(request)
+
+        await MainActor.run {
+            if let devices = response.value.items {
+                for device in devices {
+                    guard let id = device.id else { continue }
+
+                    if let existingDevice = self.devices[id] {
+                        existingDevice.value = device
+                    } else {
+                        self.devices[id] = BindingBox<DeviceInfo?>(
+                            source: .init(get: { device }, set: { _ in })
+                        )
+                    }
+                }
+
+                self.devices.sort { x, y in
+                    let device0 = x.value.value
+                    let device1 = y.value.value
+                    return (device0?.dateLastActivity ?? Date()) > (device1?.dateLastActivity ?? Date())
+                }
+            }
+        }
+    }
+
+    // MARK: - Set Custom Name
+
+    private func setCustomName(id: String, newName: String) async throws {
+        let request = Paths.updateDeviceOptions(id: id, DeviceOptionsDto(customName: newName))
+        try await userSession.client.send(request)
+
+        if let device = self.devices[id]?.value {
+            await MainActor.run {
+                self.devices[id]?.value?.name = newName
+            }
+        }
+    }
+
+    // MARK: - Delete Device
+
+    private func deleteDevice(id: String) async throws {
+        // Don't allow self-deletion
+        guard id != userSession.client.configuration.deviceID else {
+            return
+        }
+
+        let request = Paths.deleteDevice(id: id)
+        try await userSession.client.send(request)
+
+        await MainActor.run {
+            self.devices.removeValue(forKey: id)
+        }
+    }
+
+    // MARK: - Delete All Devices
+
+    private func deleteAllDevices() async throws {
+        let deviceIdsToDelete = self.devices.keys.filter { $0 != userSession.client.configuration.deviceID }
+
+        for deviceId in deviceIdsToDelete {
+            try await deleteDevice(id: deviceId)
```

Comment:
I can switch that out!

What do you think of replacing Delete All with a multi-select? Since we're just passing a list into a single delete endpoint, I was thinking like what email has:

![image](https://github.com/user-attachments/assets/fae3ef50-57b8-4222-967f-43aecca9231f)

Then someone can select all > delete if they want to OR just delete specific ones instead of doing swipe, confirm, swipe, confirm... over and over.

Reply:
Hm... It might be too crowded to do something like this:
<img width="346" alt="Screenshot 2024-10-16 at 10 11 38" src="https://github.com/user-attachments/assets/24a1cec7-8fc3-4608-9ab1-4fcb49d53162">

I've updated all the ViewModel logic so the deleteDevices takes an array of ids then, DeleteAll just passes the self.devices in. So, if we want to go this route it's an easy functional change but UI-wise I'm struggling to get something together

Reply:
I agree that sounds like a good idea and good for at least having the initial functionality. I will also have to take a look at the button layouts.

Reply:
Sounds good. I fully got rid of Delete all in favor of:

`viewModel.send(.deleteDevices(ids: Array(viewModel.devices.keys)))`

The result is the same but it should make switching over to a multi-select really straight forward from a ViewModel perspective.

I wasn't sure if deleteDevice and deleteDevices was really necessary to have both. They're both in this most recent commit and it's all working as expected.

Reply:
This version works but the animations when going into editMode are a little jarring.

Having the **Select All** / **Remove All** at the top of the List is also not the easiest UX. Unfortunately, there just isn't enough toolbar room to fit everything.

<img width="1005" alt="Screenshot 2024-10-16 at 15 29 07" src="https://github.com/user-attachments/assets/b852025e-380e-4bb2-93d5-e57b6b3e5c66">

Reply:
This is what I have now:

<img width="350" alt="Screenshot 2024-10-16 at 18 55 35" src="https://github.com/user-attachments/assets/88974462-8045-4dbb-b8c0-5185ae7037e7">

<img width="344" alt="Screenshot 2024-10-16 at 18 57 18" src="https://github.com/user-attachments/assets/1d5a9b8f-91c2-4337-8109-5a2b97cf5a65">

<img width="343" alt="Screenshot 2024-10-16 at 18 57 25" src="https://github.com/user-attachments/assets/4356953b-5311-4e81-a9ab-5d9fc4066a42">

Reply:
Delete Devices might be a little much...

Reply:
Sorry, if you didn't see my edit: I meant hide the back bar button, not the bar. We can have the large delete button overlaid on the bottom like the user selection screen.

Reply:
Apologies, I think the screenshot looked weird. Those are on the bar:

```
        contentView
            .navigationTitle(L10n.allDevices)
            .navigationBarTitleDisplayMode(.inline)
            .navigationBarBackButtonHidden(selectMode)
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    if selectMode {
                        navigationBarSelectView
                    }
                }
                ToolbarItem(placement: .principal) {
                    if selectMode {
                        navigationBarDeleteView
                    }
                }
                ToolbarItem(placement: .navigationBarTrailing) {
                    navigationBarEditView
                }
            }
```

I've committed my current working version. At least, I *think* we're on the same page with what I did haha

Reply:
> Sorry, if you didn't see my edit: I meant hide the back bar button, not the bar. We can have the large delete button overlaid on the bottom like the user selection screen.

I need to replace this delete button let me re-look at this again today and try making this closer to SelectUserView

Reply:
Okay, apologies on this one. I see what you meant about the delete button. I've reworked my deviceRow to almost be a complete copy of the userRow from that selectUserView. I think there is only one 'issue' that's purely cosmetic. The TextPair doesn't let me darken the text. I can recreate this for this struct only and allow me to darken it but I wanted to make sure with you it was worth this effort:

<img width="340" alt="Screenshot 2024-10-17 at 13 06 21" src="https://github.com/user-attachments/assets/572694cd-7911-422c-9330-e3f78d0620dd">

<img width="345" alt="Screenshot 2024-10-17 at 13 07 41" src="https://github.com/user-attachments/assets/7423f354-fb68-4db0-9f45-71036837c199">

Other than that, this should almost exactly mirror selectUserView:

<details>
<summary>Screenshots</summary>
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 44](https://github.com/user-attachments/assets/bc01933c-6f4a-4b3a-85d9-cee7d7172cd5)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 46](https://github.com/user-attachments/assets/6a0fbd45-ec8e-4335-8872-a56808cc4eb8)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 49](https://github.com/user-attachments/assets/c6f74c20-3ff1-468d-be9e-8860e3f120e6)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 53](https://github.com/user-attachments/assets/06d2464c-48bf-4360-ab31-212fba201d10)
<\details>

Reply:
`TextPairView` was using `foregroundColor`, not `foregroundStyle`. I've pushed those changes.